quic/s2n-quic-platform/src/socket/io/tx.rs

```diff
@@ -0,0 +1,439 @@
+    capacity: usize,
+    is_full: &'a mut bool,
+}
+
+impl<'a, T: Message> TxQueue<'a, T> {
+    /// Tries to send a message as a GSO segment
+    ///
+    /// Returns the Err(Message) if it was not able to. Otherwise, the index of the GSO'd message is returned.
+    #[inline]
+    fn try_gso<M: tx::Message<Handle = T::Handle>>(
+        &mut self,
+        mut message: M,
+    ) -> Result<Result<tx::Outcome, M>, tx::Error> {
+        // the message doesn't support GSO to return it
+        if !T::SUPPORTS_GSO {
+            return Ok(Err(message));
+        }
+
+        let max_segments = self.max_segments;
+
+        let (prev_message, gso) = if let Some(gso) = self.gso_message() {
+            gso
+        } else {
+            return Ok(Err(message));
+        };
+
+        debug_assert!(
+            max_segments > 1,
+            "gso_segment should only be set when max_gso > 1"
+        );
+
+        // check to make sure the message can be GSO'd and can be included in the same
+        // GSO payload as the previous message
+        let can_gso = message.can_gso(gso.size, gso.count)
+            && message.path_handle().strict_eq(&gso.handle)
+            && message.ecn() == gso.ecn;
+
+        // if we can't use GSO then flush the current message
+        if !can_gso {
+            self.flush_gso();
+            return Ok(Err(message));
+        }
+
+        debug_assert!(
+            gso.count < max_segments,
+            "{} cannot exceed {}",
+            gso.count,
+            max_segments
+        );
+
+        let payload_len = prev_message.payload_len();
+
+        unsafe {
+            // Safety: all payloads should have enough capacity to extend max_segments *
+            // gso.size
+            prev_message.set_payload_len(payload_len + gso.size);
+        }
+
+        // allow the message to write up to `gso.size` bytes
+        let buffer = &mut T::payload_mut(prev_message)[payload_len..];
+        let buffer = tx::PayloadBuffer::new(buffer);
+
+        match message.write_payload(buffer, gso.count).and_then(|size| {
+            // we don't want to send empty packets
+            if size == 0 {
+                Err(tx::Error::EmptyPayload)
+            } else {
+                Ok(size)
+            }
+        }) {
+            Err(err) => {
+                unsafe {
+                    // revert the len to what it was before
+                    prev_message.set_payload_len(payload_len);
+                }
+                Err(err)
+            }
+            Ok(size) => {
+                debug_assert_ne!(size, 0, "payloads should never be empty");
+
+                unsafe {
+                    debug_assert!(
+                        gso.size >= size,
+                        "the payload tried to write more than available"
+                    );
+                    // set the len to the actual amount written to the payload
+                    prev_message.set_payload_len(payload_len + size.min(gso.size));
```

Comment:
well the idea is to catch the error in debug but still guard against it without debug assertions and not panic, since the worse case is we truncate an outgoing packet.

Reply:
I see, okay.